tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py

```diff
@@ -0,0 +1,415 @@
+        )
+        async def test_show_progress(
+            self,
+            mock_progress_indicator,
+            mock_native_model: Mock,
+            test_case: LLMTestCase,
+            fact_classification_result: FactClassificationResult,
+        ):
+            mock_native_model.a_generate = AsyncMock(
+                return_value=(fact_classification_result, 0.5)
+            )
+
+            metric = FactualCorrectnessMetric(model=mock_native_model)
+
+            # since we patched metric_progress_indicator, it shuold call the mocked context manager mock_progress_indicator
+            await metric.a_measure(test_case, _show_indicator=True)
+
+            # test it actually called the mock
+            mock_progress_indicator.assert_called_once_with(
+                metric, async_mode=metric.async_mode, _show_indicator=True
+            )
+
+        @pytest.mark.asyncio
+        @patch(
+            "govuk_chat_evaluation.rag_answers.custom_deepeval.metrics.factual_correctness.factual_correctness.logging.debug"
+        )
+        async def test_logs_confusion_matrix(
+            self, mock_logging_debug: MagicMock, mock_native_model: Mock, test_case
+        ):
+            # confusion matrix for logging
+            sample_confusion_matrix = ClassifiedFacts(
+                TP=["fact1"], FP=["fact2"], FN=["fact3"]
+            )
+
+            metric = FactualCorrectnessMetric(model=mock_native_model)
+
+            # mock _a_classify_statements to return the sample confusion matrix
+            metric._a_classify_statements = AsyncMock(
+                return_value=sample_confusion_matrix
+            )
+
+            await metric.a_measure(test_case)
+
+            # verify logging.debug was called once
+            mock_logging_debug.assert_called_once()
+
+            # verify ligged message contains the expected objects
+            logged_message = mock_logging_debug.call_args[0][0]
+
+            assert "Confusion matrix for test input:" in logged_message
+            assert "fact1" in logged_message
+            assert "fact2" in logged_message
+            assert "fact3" in logged_message
+
+            assert str(test_case.input) in logged_message
+
+        @pytest.mark.asyncio
+        async def test_returns_score(
+            self,
+            mock_native_model: Mock,
+            test_case: LLMTestCase,
+            sample_classified_facts: ClassifiedFacts,
+        ):
+            metric = FactualCorrectnessMetric(
+                model=mock_native_model, threshold=0.7, include_reason=True
+            )
+
+            metric._a_classify_statements = AsyncMock(
+                return_value=sample_classified_facts
+            )
+
+            # let _calculate_score run naturally; it uses mocked _classify_statements
+            result = await metric.a_measure(test_case)
+
+            # expected score based on real implementation
+            expected_score = metric._calculate_score()
```

Comment:
We should use a hardcoded number here rather than asserting based off the private method. Reasons to do so are: it's easier to reason with an actual number, it makes it clearer that it actually returns an expected type, as the code itself is already using the private method it conceals issues with that method.

Reply:
Done as part of https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/f16d1789fc5e0c158f51aad2d711c4071af934af (also parametrised to test the zero if `tp+fp < 0` behaviour.